db/db_compaction_test.cc

```diff
@@ -3583,6 +3597,54 @@ TEST_P(DBCompactionWaitForCompactTest,
   ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->ClearAllCallBacks();
 }
 
+TEST_P(DBCompactionWaitForCompactTest, WaitForCompactToTimeout) {
+  // When timeout_micros is set, this test makes CompactionJob hangs forever
+  // using sync point. This test also sets the timeout to be 1 second for
+  // WaitForCompact to time out early. WaitForCompact() is expected to return
+  // Status::TimedOut.
+  // When timeout_micros is not set, we expect WaitForCompact() to wait
+  // indefinitely. We don't want the test to hang forever. When timeout_micros =
+  // 0, this test is not much different from
+  // WaitForCompactWaitsOnCompactionToFinish
+
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->LoadDependency(
+      {{"DBImpl::WaitForCompact:StartWaiting",
+        "DBCompactionTest::WaitForCompactToTimeout:0"},
+       {"DBCompactionTest::WaitForCompactToTimeout:1",
+        "CompactionJob::Run():Start"}});
+
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
+
+  // Now trigger L0 compaction by adding a file
+  Random rnd(123);
+  GenerateNewRandomFile(&rnd, /* nowait */ true);
+  ASSERT_OK(Flush());
+
+  // Wait for Compaction in another thread
+  auto waiting_for_compaction_thread = port::Thread([this]() {
+    if (wait_for_compact_options_.timeout.count()) {
+      // Make timeout shorter to finish test early
+      wait_for_compact_options_.timeout = std::chrono::microseconds{1000000};
```

Comment:
Not ideal, but OK. And might as well make the sleep very short (e.g. 1 millisecond) if using real time.

We do have provisions for mock sleeping. See SpecialEnv and example in DBOptionsTest.DeleteObsoleteFilesPeriodChange. It's preferred in general but probably not yet necessary here.

Reply:
Unfortunately, MockSleep doesn't work well with `CondVar::TimedWait()` (asked in the group chat and Andrew confirmed) so I had to go this route.